Make same origin i-frames load in the same process with site-isolation

chirags27 opened the pull request:
#### 4d2fb292a80797dbc2f76aeffa5e4b1728fe0a75
<pre>
Make same origin i-frames load in the same process with site-isolation
<a href="https://bugs.webkit.org/show_bug.cgi?id=255070">https://bugs.webkit.org/show_bug.cgi?id=255070</a>
rdar://107618125

Reviewed by Alex Christensen.

This change makes it so that we don&apos;t load each iframe in a new process,
but club together the same origin iframes in the same webcontent
process.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/LocalFrameCreationParameters.h: Copied from Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h.
* Source/WebKit/Shared/LocalFrameCreationParameters.serialization.in: Added.
* Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
(WebKit::ProvisionalFrameProxy::ProvisionalFrameProxy):
(WebKit::ProvisionalFrameProxy::~ProvisionalFrameProxy):
(WebKit::ProvisionalFrameProxy::didReceiveMessage):
(WebKit::ProvisionalFrameProxy::decidePolicyForResponse): Deleted.
(WebKit::ProvisionalFrameProxy::didCommitLoadForFrame): Deleted.
* Source/WebKit/UIProcess/ProvisionalFrameProxy.h:
(): Deleted.
* Source/WebKit/UIProcess/SubframePageProxy.cpp:
(WebKit::SubframePageProxy::SubframePageProxy):
(WebKit::SubframePageProxy::~SubframePageProxy):
(WebKit::SubframePageProxy::didReceiveMessage):
(WebKit::SubframePageProxy::decidePolicyForResponse):
(WebKit::SubframePageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/SubframePageProxy.h:
(WebKit::SubframePageProxy::process):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::~WebFrameProxy):
(WebKit::WebFrameProxy::didFinishLoad):
(WebKit::WebFrameProxy::swapToProcess):
(WebKit::WebFrameProxy::commitProvisionalFrame):
(WebKit::WebFrameProxy::updateRemoteFrameSize): Deleted.
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::updateRemoteFrameSize):
(WebKit::WebPageProxy::subframePageProxyForFrameID const):
(WebKit::WebPageProxy::didFinishLoadForFrame):
(WebKit::WebPageProxy::addSubframePageProxyForFrameID):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::WebFrameLoaderClient):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::createLocalSubframeHostedInAnotherProcess):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::constructFrameTree):
(WebKit::WebPage::loadRequestByCreatingNewLocalFrameOrConvertingRemoteFrame):
(WebKit::WebPage::loadRequest):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/262943@main">https://commits.webkit.org/262943@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9df14737387a106f91cb7cffee7d974222367f5c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3098 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3159 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3267 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4510 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3464 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3067 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3223 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3187 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2696 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3129 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3455 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2782 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4311 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/953 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2759 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2597 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2727 "4 api tests failed or timed out") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2810 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4047 "260 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3178 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2532 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2765 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2761 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2756 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/360 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3019 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/WebProcessPool.h:
```diff
@@ -452,7 +452,7 @@ class WebProcessPool final
     bool hasBackgroundWebProcesses() const { return m_backgroundWebProcessCounter.value(); }
 
-    void processForNavigation(WebPageProxy&, const API::Navigation&, Ref<WebProcessProxy>&& sourceProcess, const URL& sourceURL, ProcessSwapRequestedByClient, WebProcessProxy::LockdownMode, const FrameInfoData&, Ref<WebsiteDataStore>&&, CompletionHandler<void(Ref<WebProcessProxy>&&, SuspendedPageProxy*, const String&)>&&);
+    void processForNavigation(WebPageProxy&, const API::Navigation&, Ref<WebProcessProxy>&& sourceProcess, const URL& sourceURL, ProcessSwapRequestedByClient, WebProcessProxy::LockdownMode, const FrameInfoData&, Ref<WebsiteDataStore>&&, CompletionHandler<void(Ref<WebProcessProxy>&&, SuspendedPageProxy*, const String&, bool)>&&);
```
bool should be an enum class Something : bool so we can tell what it is supposed to mean.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -568,4 +568,38 @@ HTTPServer server({
+
+    __block bool done { false };
+    [webView _frames:^(_WKFrameTreeNode *mainFrame) {
+        EXPECT_EQ(mainFrame.childFrames.count, 1u);
```
2?

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -568,4 +568,38 @@ HTTPServer server({
+    HTTPServer server({
+        { "/example"_s, { "<iframe id='webkit_frame_1' src='https://webkit.org/a'></iframe><iframe id='webkit_frame_2' src='https://webkit.org/b'></iframe>"_s } },
+        { "/a"_s, { "<script></script>"_s } },
+        { "/b"_s, { "<script>alert('doneb')</script>"_s } }
```
We can't assume that the second iframe will load after the first

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -568,4 +568,38 @@ HTTPServer server({
+    auto webView = adoptNS([[WKWebView alloc] initWithFrame:CGRectZero configuration:configuration]);
+    webView.get().navigationDelegate = navigationDelegate.get();
+    [webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:@"https://example.com/example"]]];
+    EXPECT_WK_STREQ([webView _test_waitForAlert], "doneb");
```
Keep the alerts in an array and spin the run loop until we have 2.

achristensen07 commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -1809,8 +1810,18 @@ void WebProcessPool::removeProcessFromOriginCacheSet(WebProcessProxy& process)
+
+    auto registrableDomain = RegistrableDomain { navigation.currentRequest().url() };
+    WTFLogAlways("Chirag -- finding WebProcessPool::processForNavigation for %s", registrableDomain.string().ascii().data());
+    auto* subframePageProxy = page.subpageFrameProxyForRegistrableDomain(registrableDomain);
```
We should probably pass enough parameters here to do this only if site isolation is enabled, and only if this is a navigation of a subframe.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2135,9 +2137,13 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 
 #if ENABLE(IMAGE_ANALYSIS) && PLATFORM(MAC)
-    WKQuickLookPreviewController *quickLookPreviewController() const { return m_quickLookPreviewController.get(); }
+    WKQuickLookPreviewController* quickLookPreviewController() const { return m_quickLookPreviewController.get(); }
```
This change is incorrect.  Our style prefers a space before * for ObjC types.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -12278,6 +12308,12 @@ void WebPageProxy::generateTestReport(const String& message, const String& group
 
+void WebPageProxy::addSubframePageProxyForFrameID(WebCore::FrameIdentifier frameID, WebCore::RegistrableDomain domain, std::unique_ptr<SubframePageProxy>&& subframePageProxy)
+{
+    m_subframePageProxyForDomain.add(domain, WTFMove(subframePageProxy));
```
We will eventually need a corresponding remove call.  Doesn't need to be in this PR, but worth adding a FIXME comment for both add calls.

shivamidow commented on Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
```diff
@@ -57,94 +58,67 @@ ProvisionalFrameProxy::ProvisionalFrameProxy(WebFrameProxy& frame, Ref<WebProces
-
-    m_process->addMessageReceiver(Messages::WebPageProxy::messageReceiverName(), m_pageID, *this);
-
+    // Its possible that this is called with a WebFrameProxy which has the message already registered (SiteIsolation.IframeNavigatesSelfWithoutChangingOrigin)
```
Nit: Its -> It's

shivamidow commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -646,6 +647,12 @@ WebPageProxy::~WebPageProxy()
+    //     auto* subframePageProxy = iter.value.get();
+    //     if (auto* webFrameProxy = subframePageProxy->frame())
+    //         subframePageProxy->process().removeFrameWithRemoteFrameProcess(*webFrameProxy);
+    // }
```
Cleanup mistake?

shivamidow commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3970,9 +3977,17 @@ void WebPageProxy::continueNavigationInNewProcess(API::Navigation& navigation, W
+        frame.swapToProcess(WTFMove(newProcess), navigation.currentRequest(), didCreateNewProcess);
         return;
-    }
+    } 
```
Nit: a tailing space

shivamidow commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3822,7 +3829,7 @@ void WebPageProxy::receivedNavigationPolicyDecision(PolicyAction policyAction, A
         } else
             WEBPAGEPROXY_RELEASE_LOG(ProcessSwapping, "decidePolicyForNavigationAction: keep using process %i for navigation, reason=%" PUBLIC_LOG_STRING, processIdentifier(), reason.utf8().data());
-
+        
```
Nit: Empty spaces

shivamidow commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -1809,8 +1810,19 @@ void WebProcessPool::removeProcessFromOriginCacheSet(WebProcessProxy& process)
+        auto registrableDomain = RegistrableDomain { navigation.currentRequest().url() };
+        if (!registrableDomain.isEmpty()) {
+            auto* subframePageProxy = page.subpageFrameProxyForRegistrableDomain(registrableDomain);
+            if (subframePageProxy) {
```
`if (auto* subframePageProxy = page.subpageFrameProxyForRegistrableDomain(registrableDomain))`

shivamidow commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1776,12 +1776,50 @@ void WebPage::platformDidReceiveLoadParameters(const LoadParameters& loadParamet
+    if (frame && frame->coreFrame())
+        return;
+
+    // If there is an existing RemoteFrame, lets remove it and later convert it to a LocalFrame
```
Nit: lets -> let's

achristensen07 commented on Source/WebKit/Shared/LoadParameters.cpp:
```diff
@@ -39,6 +39,8 @@ void LoadParameters::encode(IPC::Encoder& encoder) const
     encoder << navigationID;
     encoder << frameIdentifier;
+    encoder << parentFrameIdentifier;
+    encoder << layerHostingContextIdentifier;
```
These probably shouldn't be here.

chirags27 replied:
we need these when we call `createLocalSubframeHostedInAnotherProcess` in the web process. These are used so that we can create a LocalFrame in a web process which currently has a RemoteFrame.

achristensen07 commented on Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
```diff
@@ -57,94 +58,67 @@ ProvisionalFrameProxy::ProvisionalFrameProxy(WebFrameProxy& frame, Ref<WebProces
-    m_process->addMessageReceiver(Messages::WebPageProxy::messageReceiverName(), m_pageID, *this);
-
+    // Its possible that this is called with a WebFrameProxy which has the message already registered (SiteIsolation.IframeNavigatesSelfWithoutChangingOrigin)
+    if (frame.processIdentifier() == m_process->processIdentifier()) {
```
This check should be removed.  The ProvisionalFrameProxy constructor probably shouldn't have a WebFrameProxy or a request.

achristensen07 replied:
That should allow you to also remove bool newProcess.

chirags27 replied:
It's not clear how this condition can be removed -- this is needed because `SiteIsolation.IframeNavigatesSelfWithoutChangingOrigin` runs into the case where the same WebFrameProxy is used to create a new ProvisionalFrameProxy. If we remove this, what will register for these messages from the web process? We could have SubframePageProxy do that, but that won't have a 1-1 mapping.

achristensen07 commented on Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
```diff
@@ -57,94 +58,67 @@ ProvisionalFrameProxy::ProvisionalFrameProxy(WebFrameProxy& frame, Ref<WebProces
+        drawingArea->attachToProvisionalFrameProcess(m_process);
+    }
 
     LoadParameters loadParameters;
```
This needs to be moved to a separate function, not the constructor.

chirags27 replied:
based on offline convo, we can clean this up later if we think of a different design.

achristensen07 commented on Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
```diff
@@ -57,94 +58,67 @@ ProvisionalFrameProxy::ProvisionalFrameProxy(WebFrameProxy& frame, Ref<WebProces
     loadParameters.frameIdentifier = frame.frameID();
+    if (frame.parentFrame())
+        loadParameters.parentFrameIdentifier = frame.parentFrame()->frameID();
+    loadParameters.layerHostingContextIdentifier = m_layerHostingContextIdentifier;
```
Remove.  Also remove parentFrameIdentifier here.

achristensen07 commented on Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
```diff
@@ -57,94 +58,67 @@ ProvisionalFrameProxy::ProvisionalFrameProxy(WebFrameProxy& frame, Ref<WebProces
+            process->send(Messages::WebPage::LoadRequest(loadParameters), pageID);
+        });
+    } else {
+        m_process->send(Messages::WebPage::LoadRequestByCreatingNewLocalFrameOrConvertingRemoteFrame(loadParameters), m_pageID);
```
This needs to not be in the constructor.

chirags27 replied:
based on offline convo, we can clean this up later if we think of a different design.

achristensen07 commented on Source/WebKit/UIProcess/SubframePageProxy.cpp:
```diff
@@ -59,14 +58,43 @@ uint64_t SubframePageProxy::messageSenderDestinationID() const
 void SubframePageProxy::didReceiveMessage(IPC::Connection& connection, IPC::Decoder& decoder)
 {
+#if HAVE(VISIBILITY_PROPAGATION_VIEW)
+    // FIXME: This needs to be handled correctly in a way that doesn't cause assertions or crashes..
```
nit: extra period.

achristensen07 commented on Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
```diff
@@ -57,94 +58,67 @@ ProvisionalFrameProxy::ProvisionalFrameProxy(WebFrameProxy& frame, Ref<WebProces
-
-    drawingArea->attachToProvisionalFrameProcess(m_process);
+    if (newProcess) {
+        auto parameters = page.creationParameters(m_process, *drawingArea);
```
This should only be done in SubframePageProxy constructor.

chirags27 replied:
based on offline convo, we can clean this up later if we think of a different design.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -646,6 +647,12 @@ WebPageProxy::~WebPageProxy()
         GroupActivitiesSessionNotifier::sharedNotifier().removeWebPage(*this);
 #endif
+
+    // for (auto& iter : m_subframePageProxyForDomain) {
```
for (auto& subframePageProxy : m_subframePageProxyForDomain.values())

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -3231,6 +3238,9 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 
     RefPtr<WebPageProxy> m_pageToCloneSessionStorageFrom;
+
+    HashMap<WebCore::RegistrableDomain, std::unique_ptr<SubframePageProxy>>  m_subframePageProxyForDomain;
```
Can the value be a UniqueRef instead of a unique_ptr?

achristensen07 commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -970,8 +970,10 @@ void WebProcess::didClose(IPC::Connection& connection)
-{
-    return m_frameMap.get(frameID).get();
+{   
+    if (m_frameMap.contains(frameID))
```
Revert this change, I think.  Or ASSERT(contains)

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1779,12 +1779,50 @@ void WebPage::platformDidReceiveLoadParameters(const LoadParameters& loadParamet
 
     RefPtr frame = loadParameters.frameIdentifier ? WebProcess::singleton().webFrame(*loadParameters.frameIdentifier) : m_mainFrame.ptr();
-    if (!frame) {
+    if (!frame || frame->coreRemoteFrame()) {
```
is<RemoteFrame>(frame->coreAbstractFrame())

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -150,18 +150,20 @@ Ref<WebFrame> WebFrame::createSubframe(WebPage& page, WebFrame& parent, const At
     ASSERT(!frame->m_frameID);
     frame->m_frameID = coreFrame->frameID();
     WebProcess::singleton().addWebFrame(frameID, frame.ptr());
+    if (invalidator)
```
This is a use-after-move.  I think we can just remove this whole thing.

chirags27 replied:
It looks like we might need this because we're creating a new WebFrame as well. I'll fix the use-after-move

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2705,7 +2712,7 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     void reportPageLoadResult(const WebCore::ResourceError&);
 
-    void continueNavigationInNewProcess(API::Navigation&, WebFrameProxy&, std::unique_ptr<SuspendedPageProxy>&&, Ref<WebProcessProxy>&&, ProcessSwapRequestedByClient, WebCore::ShouldTreatAsContinuingLoad, std::optional<NetworkResourceLoadIdentifier> existingNetworkResourceLoadIdentifierToResume);
+    void continueNavigationInNewProcess(API::Navigation&, WebFrameProxy&, std::unique_ptr<SuspendedPageProxy>&&, Ref<WebProcessProxy>&&, ProcessSwapRequestedByClient, WebCore::ShouldTreatAsContinuingLoad, std::optional<NetworkResourceLoadIdentifier> existingNetworkResourceLoadIdentifierToResume, bool = true);
```
bool should have clearer name.

achristensen07 commented on Source/WebKit/UIProcess/WebFrameProxy.h:
```diff
@@ -152,7 +151,7 @@ class WebFrameProxy : public API::ObjectImpl<API::Object::Type::Frame>, public I
     void didCreateSubframe(WebCore::FrameIdentifier);
     ProcessID processIdentifier() const;
-    void swapToProcess(Ref<WebProcessProxy>&&, const WebCore::ResourceRequest&);
+    void swapToProcess(Ref<WebProcessProxy>&&, const WebCore::ResourceRequest&, bool);
```
give bool a name.

achristensen07 commented on Source/WebKit/UIProcess/SubframePageProxy.cpp:
```diff
@@ -33,18 +33,20 @@
 namespace WebKit {
 
-SubframePageProxy::SubframePageProxy(WebFrameProxy& frame, WebPageProxy& page, WebProcessProxy& process)
+SubframePageProxy::SubframePageProxy(WebPageProxy& page, WebProcessProxy& process, bool isMainFrame)
```
isMainFrame should be isInSameProcessAsMainFrame

cdumez commented on Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
```diff
@@ -47,7 +49,7 @@
 namespace WebKit {
 
-ProvisionalFrameProxy::ProvisionalFrameProxy(WebFrameProxy& frame, Ref<WebProcessProxy>&& process, const WebCore::ResourceRequest& request)
+ProvisionalFrameProxy::ProvisionalFrameProxy(WebFrameProxy& frame, Ref<WebProcessProxy>&& process, const WebCore::ResourceRequest& request, bool newProcess)
```
newProcess is a bad name for a boolean. Per WebKit coding style, boolean variables/parameters need a prefix (e.g. shouldUseNewProcess).

chirags27 committed:
Make same origin i-frames load in the same process with site-isolation
https://bugs.webkit.org/show_bug.cgi?id=255070
rdar://107618125

Reviewed by Alex Christensen.

This change makes it so that we don't load each iframe in a new process,
but club together the same origin iframes in the same webcontent
process.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/LocalFrameCreationParameters.h: Copied from Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h.
* Source/WebKit/Shared/LocalFrameCreationParameters.serialization.in: Added.
* Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
(WebKit::ProvisionalFrameProxy::ProvisionalFrameProxy):
(WebKit::ProvisionalFrameProxy::~ProvisionalFrameProxy):
(WebKit::ProvisionalFrameProxy::didReceiveMessage):
(WebKit::ProvisionalFrameProxy::decidePolicyForResponse): Deleted.
(WebKit::ProvisionalFrameProxy::didCommitLoadForFrame): Deleted.
* Source/WebKit/UIProcess/ProvisionalFrameProxy.h:
(): Deleted.
* Source/WebKit/UIProcess/SubframePageProxy.cpp:
(WebKit::SubframePageProxy::SubframePageProxy):
(WebKit::SubframePageProxy::~SubframePageProxy):
(WebKit::SubframePageProxy::didReceiveMessage):
(WebKit::SubframePageProxy::decidePolicyForResponse):
(WebKit::SubframePageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/SubframePageProxy.h:
(WebKit::SubframePageProxy::process):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::~WebFrameProxy):
(WebKit::WebFrameProxy::didFinishLoad):
(WebKit::WebFrameProxy::swapToProcess):
(WebKit::WebFrameProxy::commitProvisionalFrame):
(WebKit::WebFrameProxy::updateRemoteFrameSize): Deleted.
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::updateRemoteFrameSize):
(WebKit::WebPageProxy::subframePageProxyForFrameID const):
(WebKit::WebPageProxy::didFinishLoadForFrame):
(WebKit::WebPageProxy::addSubframePageProxyForFrameID):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::WebFrameLoaderClient):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::createLocalSubframeHostedInAnotherProcess):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::constructFrameTree):
(WebKit::WebPage::loadRequestByCreatingNewLocalFrameOrConvertingRemoteFrame):
(WebKit::WebPage::loadRequest):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/262943@main

cdumez commented on Source/WebKit/UIProcess/ProvisionalFrameProxy.h:
```diff
@@ -52,7 +52,7 @@ struct FrameInfoData;
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    ProvisionalFrameProxy(WebFrameProxy&, Ref<WebProcessProxy>&&, const WebCore::ResourceRequest&);
+    ProvisionalFrameProxy(WebFrameProxy&, Ref<WebProcessProxy>&&, const WebCore::ResourceRequest&, bool);
```
We should NOT omit the boolean parameter name here as it is not obvious.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5563,6 +5585,13 @@ void WebPageProxy::didFinishDocumentLoadForFrame(FrameIdentifier frameID, uint64
 
+SubframePageProxy* WebPageProxy::subframePageProxyForFrameID(WebCore::FrameIdentifier identifier) const
+{
+    if (m_registrableDomainForSubframeID.contains(identifier))
```
We're doing a double hash map lookup in m_registrableDomainForSubframeID here and on the next line. This is unnecessarily inefficient.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -3231,6 +3238,9 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 
     RefPtr<WebPageProxy> m_pageToCloneSessionStorageFrom;
+
+    HashMap<WebCore::RegistrableDomain, UniqueRef<SubframePageProxy>>  m_subframePageProxyForDomain;
```
I don't think these HashMaps have good names. It is a container but it sounds like a single SubframePageProxy.
Also you have an extra space between the type and the name.

For the name, I could see something like m_domainToSubframePageProxyMap.

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.h:
```diff
@@ -450,7 +450,11 @@ class WebProcessPool final
     bool hasBackgroundWebProcesses() const { return m_backgroundWebProcessCounter.value(); }
 
-    void processForNavigation(WebPageProxy&, const API::Navigation&, Ref<WebProcessProxy>&& sourceProcess, const URL& sourceURL, ProcessSwapRequestedByClient, WebProcessProxy::LockdownMode, const FrameInfoData&, Ref<WebsiteDataStore>&&, CompletionHandler<void(Ref<WebProcessProxy>&&, SuspendedPageProxy*, ASCIILiteral)>&&);
+    enum class DidCreateNewProcess : bool {
```
We usually have these on a single line.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1779,12 +1780,55 @@ void WebPage::platformDidReceiveLoadParameters(const LoadParameters& loadParamet
+        return;
+    }
+
+    // If there is an existing RemoteFrame, let's remove it and later convert it to a LocalFrame
```
Comment is missing a period at the end (per coding style).

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1779,12 +1780,55 @@ void WebPage::platformDidReceiveLoadParameters(const LoadParameters& loadParamet
+
+        parent->tree().removeChild(*coreRemoteFrame);
+        coreRemoteFrame->disconnectOwnerElement();
+        if (auto* remoteFrameClient = static_cast<WebRemoteFrameClient*>(&coreRemoteFrame->client()))
```
We should add the right template specializations so we can use downcast.

achristensen07 replied:
I don't think so in this case.  This follows the pattern of static_cast<WebFrameLoaderClient

cdumez replied:
I don’t understand your point. Why don’t we want a safer cast here? Or why can’t we?
The fact that existing code is unsafe doesn’t sound like a good reason to not make code safer.
